woocommerce/woocommerce-gutenberg-products-block
# Add validation context provider and implement validation for shipping country and coupons.

Fixes: #1969
Fixes: #1966

Some of the work in this pull can also help with #1939 (cc @Aljullu )

As a part of our work on the checkout and cart blocks, we need a way to validate fields, show validation errors, and track validation state. In this pull I've tackled validation for two things:

- Give an indication to customer when they fill our address fields but there is no country selected.
- Give an indication to customer when invalid coupon is applied (currently this is via notice, but we can do this via inline validation to avoid scroll).

I thought these two use-cases would be involve straightforward validation logic giving opportunity to validate (!pun) an approach I had in mind for using context to help with validation.

So in this pull is the following:

- Addition `ValidationContextProvider` and `useValidationContext`. This provides state for validation errors which are stored as an object where properties are the field the validation error applies to and values are the user friendly message to show to customer's when a validation fails.
- Addition of `ValidationInputError` component. This component is used in field components wherever you want a validation error to appear. It applies a consistent style and format to the error message.
- Implemented the new `ValidationInputError` component with the `CountryInput` and also created a storybook for the `CountryInput` so that the we could see the validation at work!  Note: There is a console.error with the story whenever a country is selected. The story is still completely functional and this console error doesn't appear in actual use. I think it's a storybook bug because the error doesn't make sense (might have to do with decorators being used).

![Screen Recording 2020-03-13 at 10 09 PM](https://user-images.githubusercontent.com/1429108/76674179-d1e7ba80-6582-11ea-84cb-5f2be2cbc01e.gif)

- Implemented the new `ValidationInputError` component with the `TotalsCouponCodeInput` component and also created a story book for this.
- Various improvements in related code touched in this pull (I'll note some of them via inline comments in the code).

![Screen Recording 2020-03-14 at 06 21 PM](https://user-images.githubusercontent.com/1429108/76691386-b7641e80-6620-11ea-9bd7-de8419439dfd.gif)

## To Test

- Verify cart loads in editor and frontend without errors
- Verify checkout loads in editor and frontend without errors.
- **Country validation:** Set your site so that "Default Customer Location" is set to "No location by default" and then load an incognito browser view. In the shipping calculator select/enter in a field without selecting country. A validation error should appear and the update button should be disabled. As soon as you select country the validation error should disappear and the button re-enabled. Verify shipping rates etc work as before.
- **Coupon validation:** Coupon notices for coupon errors have been replaced by the validation error. Verify this works as expected and coupon behaviour for valid coupons works as expected too.
- start up storybook and you can see components and their behaviour in storybook as well (`CountryInput` and `TotalsCouponCodeInput`)

Observer styles etc and report on any issues or unexpected behaviour (see gifs for what is expected).
